File: debop4s-timeperiod/src/test/scala/com/github/debop4s/timeperiod/tests/timeranges/YearRangeCollectionTest.scala
package com.github.debop4s.timeperiod.tests.timeranges

import com.github.debop4s.timeperiod._
import com.github.debop4s.timeperiod.tests.AbstractTimePeriodTest
import com.github.debop4s.timeperiod.timerange.{YearRange, YearRangeCollection}
import com.github.debop4s.timeperiod.utils.Times

/**
 * com.github.debop4s.timeperiod.tests.timeranges.YearRangeCollectionTest
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 2014. 2. 17. 오전 10:41
 */
class YearRangeCollectionTest extends AbstractTimePeriodTest {

    test("single years") {
        val startYear = 2014

        val yrs = YearRangeCollection(startYear, 1)
        yrs.yearCount should equal(1)
        yrs.startYear should equal(startYear)
        yrs.endYear should equal(startYear)

        val years = yrs.years
        years.size should equal(1)
        years(0).isSamePeriod(YearRange(startYear)) should equal(true)
    }

    test("calendar months") {
        val startYear = 2014
        val yearCount = 5

        val yrs = YearRangeCollection(startYear, yearCount, EmptyOffsetTimeCalendar)
        yrs.yearCount should equal(yearCount)
        yrs.startYear should equal(startYear)
        yrs.endYear should equal(startYear + yearCount)
    }

    test("year counts") {
        val yearCounts = Array(1, 6, 12, 48, 180, 360)
        val now = Times.now
        val today = Times.today

        yearCounts.foreach { yearCount =>
            val yearRanges = YearRangeCollection(now, yearCount)
            val startTime = yearRanges.calendar.mapStart(Times.trimToYear(today))
            val endTime = yearRanges.calendar.mapEnd(startTime + yearCount.year)

            yearRanges.start should equal(startTime)
            yearRanges.end should equal(endTime)

            val items = yearRanges.years

            (0 until yearCount).par.foreach { y =>
                val item = items(y)

                item.start should equal(startTime + y.year)
                item.end should equal(yearRanges.calendar.mapEnd(startTime + (y + 1).year))

                item.unmappedStart should equal(startTime + y.year)
                item.unmappedEnd should equal(startTime + (y + 1).year)

                item.isSamePeriod(YearRange(yearRanges.start + y.year)) should equal(true)
            }
        }
    }

}

File: debop4s-timeperiod/src/test/java/debop4s/timeperiod/tests/timeranges/YearRangeCollectionTest.java
package debop4s.timeperiod.tests.timeranges;

import debop4s.core.JAction1;
import debop4s.core.parallels.JParallels;
import debop4s.timeperiod.TimeCalendar;
import debop4s.timeperiod.tests.TimePeriodTestBase;
import debop4s.timeperiod.timerange.YearRange;
import debop4s.timeperiod.timerange.YearRangeCollection;
import debop4s.timeperiod.utils.Times;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;
import org.junit.Test;

import java.util.List;

import static org.fest.assertions.Assertions.assertThat;

/**
 * debop4s.timeperiod.test.timeranges.YearRangeCollectionFunSuite
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 5. 26. 오전 12:12
 */
@Slf4j
public class YearRangeCollectionTest extends TimePeriodTestBase {
    @Test
    public void singleYears() {
        final int startYear = 2004;

        YearRangeCollection yearRanges = new YearRangeCollection(startYear, 1);
        assertThat(yearRanges.getYearCount()).isEqualTo(1);

        assertThat(yearRanges.getStartYear()).isEqualTo(startYear);
        assertThat(yearRanges.getEndYear()).isEqualTo(startYear);

        List<YearRange> years = yearRanges.getYears();
        assertThat(years.size()).isEqualTo(1);
        assertThat(years.get(0).isSamePeriod(new YearRange(startYear))).isTrue();
    }

    @Test
    public void calenarMonths() {
        final int startYear = 2004;
        final int yearCount = 5;

        YearRangeCollection yearRanges = new YearRangeCollection(startYear, yearCount, TimeCalendar.getEmptyOffset());

        assertThat(yearRanges.getYearCount()).isEqualTo(yearCount);
        assertThat(yearRanges.getStartYear()).isEqualTo(startYear);
        assertThat(yearRanges.getEndYear()).isEqualTo(startYear + yearCount);
    }

    @Test
    public void yearCounts() {
        int[] yearCounts = new int[] { 1, 6, 48, 180, 365 };

        final DateTime now = Times.now();
        final DateTime today = Times.today();

        for (int yearCount : yearCounts) {
            final YearRangeCollection yearRanges = new YearRangeCollection(now, yearCount);

            final DateTime startTime = yearRanges.getTimeCalendar().mapStart(Times.trimToYear(today));
            final DateTime endTime = yearRanges.getTimeCalendar().mapEnd(startTime.plusYears(yearCount));

            assertThat(yearRanges.getStart()).isEqualTo(startTime);
            assertThat(yearRanges.getEnd()).isEqualTo(endTime);

            final List<YearRange> items = yearRanges.getYears();

            JParallels.run(yearCount, new JAction1<Integer>() {
                @Override
                public void perform(Integer y) {
                    final YearRange item = items.get(y);
                    assertThat(item.getStart()).isEqualTo(startTime.plusYears(y));
                    assertThat(item.getEnd()).isEqualTo(yearRanges.getTimeCalendar().mapEnd(startTime.plusYears(y + 1)));

                    assertThat(item.getUnmappedStart()).isEqualTo(startTime.plusYears(y));
                    assertThat(item.getUnmappedEnd()).isEqualTo(startTime.plusYears(y + 1));

                    assertThat(item.isSamePeriod(new YearRange(yearRanges.getStart().plusYears(y)))).isTrue();
                }
            });
        }
    }
}
